tools/checked-c-convert/ProgramInfo.cpp

```diff
@@ -26,53 +26,94 @@ tyToStr(const Type *T) {
 	ConstraintVariable(ConstraintVariable::PointerVariable, 
-					   tyToStr(QT.getTypePtr())),FV(nullptr)
-{ 
+					   tyToStr(QT.getTypePtr()),N),FV(nullptr)
+{   
 	QualType QTy = QT;
 	const Type *Ty = QTy.getTypePtr();
+  // If the type is a decayed type, then maybe this is the result of 
+  // decaying an array to a pointer. If the original type is some 
+  // kind of array type, we want to use that instead. 
+  if (const DecayedType *DC = dyn_cast<DecayedType>(Ty)) {
+    QualType QTytmp = DC->getOriginalType();
+    const Type * Tytmp = QTytmp.getTypePtr();
+    if (Tytmp->isArrayType() || Tytmp->isIncompleteArrayType()) {
+      QTy = QTytmp;
+      Ty = QTy.getTypePtr();
+    }
+  }
+
 	bool isTypedef = false;
 
 	if (Ty->getAs<TypedefType>())
 		isTypedef = true;
 
-	while (Ty->isPointerType()) {
-		// Allocate a new constraint variable for this level of pointer.
-		vars.insert(K);
-		VarAtom * V = CS.getOrCreateVar(K);
-   
-    if (Ty->isCheckedPointerType()) {
-      if (Ty->isCheckedPointerPtrType()) {
-        // Constrain V so that it can't be either wild or an array.
-        CS.addConstraint(CS.createNot(CS.createEq(V, CS.getArr()))); 
-        CS.addConstraint(CS.createNot(CS.createEq(V, CS.getWild())));
-        ConstrainedVars.insert(K);
-      } else if (Ty->isCheckedPointerArrayType()) {
-        llvm_unreachable("unsupported!");
+  arrPresent = false;
+	while (Ty->isPointerType() || Ty->isArrayType()) {
+    if (Ty->isArrayType() || Ty->isIncompleteArrayType()) {
+      arrPresent = true;
+      // If it's an array, then we need both a constraint variable 
+      // for each level of the array, and a constraint variable for 
+      // values stored in the array. 
+      vars.insert(K);
+      CS.getOrCreateVar(K);
+
+      // See if there is a constant size to this array type at this position.
+      if (const ConstantArrayType *CAT = dyn_cast<ConstantArrayType>(Ty)) {
+        arrSizes[K] = std::pair<OriginalArrType,uint64_t>(
+                        O_SizedArray,CAT->getSize().getZExtValue());
+      } else {
+        arrSizes[K] = std::pair<OriginalArrType,uint64_t>(
+                        O_UnSizedArray,0);
       }
-    } 
-
-		// Save here if QTy is qualified or not into a map that 
-		// indexes K to the qualification of QTy, if any.
-		if (QTy.isConstQualified()) 
-			QualMap.insert(
-				std::pair<uint32_t, Qualification>(K, ConstQualification));
-
-		K++;
-    std::string TyName = tyToStr(Ty);
-    // TODO: Github issue #61: improve handling of types for
-    // // variable arguments.
-		if (TyName == "struct __va_list_tag *" || TyName == "va_list")
-			break;
-
-		// Iterate.
-		QTy = QTy.getSingleStepDesugaredType(C);
-		QTy = QTy.getTypePtr()->getPointeeType();
-		Ty = QTy.getTypePtr();
+
+      K++;
+
+      // Iterate.
+      // TODO: Why is it the case that we can only get the unqualified type
+      //       of an array element? That's okay, right? 
```

Comment:
No, this would cause us to drop const on array elements.  You should call getElementType() instead on the base class of ArrayType.  You need to cast the qualified type to be an ArrayType first.